### ISSUE TITLE
Primitive values

### ISSUE THREAD

**Issue by mdbs99** [mdbs99/james#28]
Let's create some classes that will represent primitives values like Integer, Currency, Text, and so on.

We should create some interfaces first like `IInteger`, `ICurrency`, etc. Thus, we can create the first basic implementations.
Another aproach is create just one interface `IScalar<T>` similiar to [this](https://github.com/yegor256/cactoos/blob/master/src/main/java/org/cactoos/Scalar.java) one, and implement it in each class.

We need discuss more about it.

**Comment by nunopicado** [comment]
@mdbs99  What's you take on the **_maybe_** construct from function languages like Haskell?
They represent a value or the lack of it.
Conceptually speaking, every value would have a Boolean value attached to it, that would represent if the main value is defined or not.
I find this very helpful when implementing _lazy initialization_.
This would also keep primitive values contained inside objects, like so:

```delphi
var
  V1: IMaybe<Integer>;
  V2: IMaybe<Word>;
begin
  V1 := TMaybe<Integer>.New;
  V2 := TMaybe<Word>.New(222);

  if not V1.IsDefined
    then V1.Define(111);  
  Result := V1.Value;

  if Value2.IsDefined
    then // do something
    else // do something else
```

The Define function, which allows defining the value after the object creation, could also be replaced by calling the New again, in order to achieve immutability;

```delphi
var
  V: IMaybe<Integer>;
begin
  V := TMaybe<Integer>.New;

  //
  //
  //

  if not V.IsDefined
    then V := TMaybe<Integer>.New(111);  
  Result := V.Value;
```

**Comment by mdbs99** [comment]
Don't you think this is a procedural approach?
I mean, what is the difference between this and testing if a variable is NULL?

**Comment by nunopicado** [comment]
The difference here is the object wouldn't be null, even when no value was present. Rather it would be an empty object (like the null object pattern, only the object itself would allow one of two states, depending on the constructor parameters).
In Haskell this is considered an optional value, which comes in handy for lazy initialization. I'm not sure if there are other ways.

**Comment by mdbs99** [comment]
Even so we need to test if the value exists... I don't think this is a good idea (if I understood right the idea).
However, we can simulate the same concept using (new) `TIf`class. What do you think?

**Comment by nunopicado** [comment]
hmmm I'm not seeing what you mean.
TIf object returns one of two values depending on the condition, and here we would need it to return only one value or none at all.
Even if we could make TIf's ValueIfFalse method optional, in case of a negative condition we wouldn't be able to return the same value the variable already had.

Can you elaborate your idea so I can see if I understood it well?

**Comment by mdbs99** [comment]
Using your code:
```pascal
  if not V1.IsDefined then
    V1.Define(TDefaultInteger.New);  
  Result := V1.Value;
```
If the "value" was not defined, I initialized it using a "default Integer" — could be a "fake Integer", "empty Integer", whatever.
The part of code that needs a Integer doesn't care about how it was initialized. The code needs the final "value", ie., an instance that can do the job properly.

So, changing the code above using TIf:
```pascal
  Result := TIf<Integer>.New(V1.IsDefined, V1.Value, TDefaultInteger.New);
```
Isn't it the same?

**Comment by nunopicado** [comment]
Actually, it's not the same.

The idea behind lazy initialization is to have a value loaded only once, but not at object creation time.
Say you want to have a database connection. You created the object, but only establish the connection upon first use.
Applying this to the concept of TMaybe, you don't want it to return a default value. you want it to return the correct value, but one that only is loaded (defined) once, and can then be accessed several times.

For instance, you have an ini file, wrapped in an object that accesses it and returns its values.
You have fields for every item in the ini file, and a function that the consumer of the class can access to get that information.
You don't want those functions to be reading the ini file every time they are called. Only once is enough.
But you don't want either to have the ini file reading inside the constructor, for obvious reasons.

You can either have another public method, in which the ini file is actually read (or delegates to a private method, unimportant at the moment how it does the reading), or you can have some kind of flag that helps you understand if the file has been read or it needs to, for example, a boolean field, stating if the file has been read or not.
Using a boolean field has some conceptual issues. If you use only one, when you call one of the functions, and the boolean field shows the file hasn't been read yet, you'll have to tell it to read the whole file so you can then retrieve only one field (the one that function is meant to access). This is a not so lazy initialization. It does wait for the first call, but at that moment, reads the whole file.

With TMaybe, you can ask the value itself if it has been read, instead of relying on an additional field, and if it has not, read only that value, and not the whole file.
It will always return the same value, but it will do an additional action (the reading of that value in the file) on the first call.

The TIf approach would not be able to do that.

**Comment by mdbs99** [comment]
But when I call this `Result := V1.Value;` inside other method from other object, I'm just expecting the `Value`. The object doesn't care if is the first reading.

My question is: Using `TMaybe` I need to use `if not V1.IsDefined then...` in my own code?

I agree about using a flag or something like `if not Assigned(FObject) then... ` is not perfect. 
So, show me some code to explain your idea better. Could you?

**Comment by nunopicado** [comment]
> But when I call this Result := V1.Value; inside other method from other object, I'm just expecting the Value. The object doesn't care if is the first reading.

Actually, the kind of object that consumes this must care about the reading, because that's his job. This is not meant for the guy who assembles the App. This is meant for the guy who creates the components that will be used to assemble the App.

Let's put it in code...

Your co-worker is developing a software that will load it's settings from an ini file.
It's your job to build the object that gets those settings, because your co-worker doesn't care when or how those settings are being read.

So you create something like this:
```delphi
  ISettings = Interface
    function Username: String;
    function Password: String;
  end;

  TSettings = Class(TInterfacedObject, ISettings)
  private
    FIni: IIniFile;
  public
    constructor Create(const IniFile: IIniFile); 
    // For the sake of brevity, let's pretend theres an interfaced version of TIniFile
    // that will be passed in in the constructor
    function Username: String;
    function Password: String;
  end;
```

On possible implementation of this would be:
```delphi
constructor TSettings.Create(const IniFile: IIniFile);
begin
  FIni := IniFile;
end;

function TSettings.Username: String;
begin
  Result := FIni.ReadString('App', 'Username', '');
end;

function TSettings.Password: String;
begin
  Result := TDecryptor.New(    // This shouldn't be created here, but it's enough for now
    FIni.ReadString('App', 'Password', '')
  ).AsString;
end;
```

Your co-worker will use this object, and nothing wrong would come of it...
Except performance.
If he calls either function more than once, it will read from disk every time.
If he calls the Password function more than once, not only it has to read from the disk multiple times, it also has to decrypt it multiple times.

That's where TMaybe can be helpfull.

If you have optional (TMaybe) fields, like this:
```delphi
  TSettings = Class(TInterfacedObject, ISettings)
  private
    FIni: IIniFile;
    FUser: IMaybe<String>;
    FPass: IMaybe<String>;
  public
    constructor Create(const IniFile: IIniFile); 
    // For the sake of brevity, let's pretend theres an interfaced version of TIniFile
    // that will be passed in in the constructor
    function Username: String;
    function Password: String;
  end;
```

And initialize those values (lazy initialization):
```delphi
constructor TSettings.Create(const IniFile: IIniFile);
begin
  FIni := IniFile;
  FUser := TMaybe<String>.New;
  FPass := TMaybe<String>.New;
end;
```

And your functions check their own value:
```delphi
function TSettings.Username: String;
begin
  if not FUser.IsDefined
    then FUser.Define(FIni.ReadString('App', 'Username', ''));
  Result := FUser.Value;
end;

function TSettings.Password: String;
begin
  if not FPass.IsDefined
    then FPass.Define(
      TDecryptor.New(    
        FIni.ReadString('App', 'Password', '')
      ).AsString
    );
  Result := FPass.Value;
end;
```

You, the TSettings creator, would make sure your code is not spending the user's time reading from disk (or whatever it is that IIniFile is doing).
And your co-worker would have a fast TSettings object from where to get his data without caring what or how it works. He just gets his values without thinking about it.

Now, I agree that If statement is not the most beautiful thing. But I can't think of another way to do lazy initialization without it.
How would you refactor this while preserving the behavior?

**Comment by mdbs99** [comment]
> Actually, the kind of object that consumes this must care about the reading, because that's his job. This is not meant for the guy who assembles the App. This is meant for the guy who creates the components that will be used to assemble the App.

No. I was talking about a `Bar` object that calls `Foo.Exec`. So, `Bar` doesn't care what's happening inside `Exec`.

> That's where TMaybe can be helpfull.
> ...
> If you have optional (TMaybe) fields, like this:
> ...
> How would you refactor this while preserving the behavior?

Now it's clear for me.
You're right, this could be a good feature, but I dont't agree this naming: `TMaybe`.

"Maybe" is more related with [`Nullable`](https://docs.microsoft.com/pt-br/dotnet/api/system.nullable-1?view=netframework-4.7) class in C#. But as I know, `Nullable` only works with primitives types, not objects. We want both.

However, this is not a good name either. What we want is a Lazy behavior, right?

What do you think about `TLaze` name? [1] 

I would not like to code using this `if not FUser.IsDefined...`
For this new object to be a truly laze value, it needs to define itself, without no IF outside to control it. I believe we can do this using anonymous method. But I don't know which is the status about this feature in FPC.

I'll do a research.

---
[1] laze is a noun

**Comment by nunopicado** [comment]
The Maybe name came directly from Haskell, from where I learned this concept and created a reproduction in Delphi. I used the same name they have over there.

TLaze: I guess I'd never think of that one, but it doesn't shock me. I'm good... 

Anonymous could be a way, yes.
~~I've used them in Lazarus a couple of years ago, nothing fancy, but needed it at the time. Didn't have problems with that. If we think about it, an event is also a sort of anonymous method, and those have been around since the start, so I'm not expecting any problems.~~
EDIT: Actually I used events, not methods. Events require a named function to pass through, should be avoided in this case.

So I was trying it out to check if it would be an option, and I really like the out come. The if is of course needed anyway, but not it sits inside the Value function of the 'Maybe/Laze' class:

```delphi
  V := TLaze<Integer>.New(
         Function: Integer
         begin
           Write('Defining value: ');
           Result := 5;
         end
       );
  WriteLn('Shall we begin?');
  WriteLn(V.Value);
  WriteLn(V.Value);
  WriteLn(V.Value);
```
Outputs this:

> Shall we begin?
> Defining value: 5
> 5
> 5

So lazy initialization is achieved (the anonymous function is called only one time, and only when the first Value function is called the first time.

Only the fact that we have to use the data type twice feels less than perfect, but I guess it still is better than the if statement.

What to you think?

**Comment by mdbs99** [comment]
> The Maybe name came directly from Haskell, from where I learned this concept and created a reproduction in Delphi. I used the same name they have over there.

Yeah, but did you just copy or really like it?

> TLaze: I guess I'd never think of that one, but it doesn't shock me. I'm good...

Actually I'm not satisfied with this name.

> What to you think?

I believe that could be a good way to try but, unfortunately, FPC still not working with anonymous functions in the currenct version.
I made some research and tried to compile some code and didn't work.

Let's think in a better name first.

**Comment by nunopicado** [comment]
> Yeah, but did you just copy or really like it?

At first, I just  copied it. Didn't really like the name, but couldn't think of a better one.
I even talked about it with a colleague (the one who showed me the function in Haskell) that the name was strange.
After using it a few times, I grew accustomed to it, and now I think I like it.
When using it with the specialized type, it makes sense:

`TMaybe<Integer>` it makes sense to an optional value.
Not so much to a lazy type, which is what it really became with the use of generics to allow it to initialize itself on demand... 

> Actually I'm not satisfied with this name.

Like I said, not one I would think of.

> I made some research and tried to compile some code and didn't work.

Yes, I found out the same.
I guess we'll have to put it on hold for now, and meanwhile we can try to came up with a better name.

**Comment by ortuagustin** [comment]
I think it's better to use a Decorator to wrap an existing ISettings to cache the return values.

**Comment by mdbs99** [comment]
@ortuagustin but the "IF's" will still exist inside this decorator instance, right?

If, for each cached result, we will need to code "IF's" (in all those classes that it's necessary), I need to say that is not acceptable.

I'm thinking in a better solution.

**Comment by mdbs99** [comment]
I'll create another issue to discuss this.

Now, please, let's come back to the issue point.

**Comment by nunopicado** [comment]
So, getting back to this issue, what pros and cons do you foresee in creating objects to represent primitive values?

**Comment by mdbs99** [comment]
The only one cons that I see is the supposed loss of performance for a few cases. But in those cases which performance is more important, we can continue using primitives.

**Comment by nunopicado** [comment]
And for the sake of clarity, can you please name a few pros as well?

**Comment by mdbs99** [comment]
Of course.
First of all, we will use objects; We can code in a declarative way; We can use decorators to convert that values but in a lazy way; We can create a `TNullable` class like C# did, but only to work in specific cases, eg, DB; We can pass those object as parameters without using procedural functions like `IntToStr` — but this is another point at the future, when we will make some special classes for that.

Java has the same: `int` <-> `Integer`

**Comment by nunopicado** [comment]
I see.
I've been using record helpers to add functionality to primitives (not that often, since Delphi already has a number of them - very useful by the way - in the SysUtils standard unit. Things like IntToStr are not really required since we can just use `i.ToString`.
But I don't know if that was already implemented in FPC.
Record helpers have the advantage of being pure primitives for all intents and purposes, while having extra functionality.
I'm not sure they would fulfill all the pros of "primitive objects", but they probably don't fit this project philosophy.

**Comment by mdbs99** [comment]
> But I don't know if that was already implemented in FPC.

Yes, that was implemented in FPC:
https://github.com/graemeg/freepascal/blob/master/rtl/objpas/sysutils/syshelph.inc
http://wiki.freepascal.org/Helper_types#record_helper
https://www.freepascal.org/docs-html/ref/refse62.html

> I'm not sure they would fulfill all the pros of "primitive objects", but they probably don't fit this project philosophy.

Unfortunately, they not.
Of course we can use them inside methods.

**Comment by mdbs99** [comment]
I've created the IScalar interface, see ->https://github.com/mdbs99/james/commit/a38c32c39ad11a31b046c67b91b21cbda0577636
